File: Burp/grabquestions.py
import subprocess
import json
import time
import hashlib
from pathlib import Path

BANK_CHOICES = [
    "Calculations",
    "Clinical Therapeutics",
    "Clinical MEP",
    "Clinical OTC",
    "Clinical Mixed"
]

WEIGHTING_CHOICES = ["Low", "Medium", "High"]

def prompt_for_bank():
    print("Select a question bank:")
    for i, name in enumerate(BANK_CHOICES, 1):
        print(f"{i}. {name}")
    while True:
        try:
            choice = int(input("Enter number: "))
            if 1 <= choice <= len(BANK_CHOICES):
                return BANK_CHOICES[choice - 1]
        except ValueError:
            pass
        print("Invalid input. Try again.")

def prompt_for_weighting():
    print("Select a weighting:")
    for i, name in enumerate(WEIGHTING_CHOICES, 1):
        print(f"{i}. {name}")
    while True:
        try:
            choice = int(input("Enter number: "))
            if 1 <= choice <= len(WEIGHTING_CHOICES):
                return WEIGHTING_CHOICES[choice - 1]
        except ValueError:
            pass
        print("Invalid input. Try again.")

# Define hash function early
def hash_question(q):
    return hashlib.sha256((q.get("title", "") + q.get("text", "")).encode()).hexdigest()

# Constants
BATCH_LIMIT = 20
SLEEP_SECONDS = 2
THRESHOLD_NEW_UNIQUES = 0.9  # If fewer than 90% of a batch are new, stop
OUTPUT_DIR = "output"

def fetch_batch(curl_command):
    result = subprocess.run(curl_command, capture_output=True, text=True)
    try:
        json_data = json.loads(result.stdout)
        return json_data
    except json.JSONDecodeError:
        print("❌ Failed to decode JSON")
        return []

def main():
    bank_name = prompt_for_bank()
    if bank_name == "Calculations":
        weighting = None
        print(f"\n📥 Scraping '{bank_name}' questions (no weighting)...\n")
        output_file = Path(OUTPUT_DIR) / f"{bank_name.lower().replace(' ', '_')}_questions.json"
        url = f"https://www.preregshortcuts.com/api/questions/get/?bank={bank_name.replace(' ', '+')}&num_questions=40"
    else:
        weighting = prompt_for_weighting()
        print(f"\n📥 Scraping '{bank_name}' questions with '{weighting}' weighting...\n")
        output_file = Path(OUTPUT_DIR) / f"{bank_name.lower().replace(' ', '_')}_{weighting.lower()}_questions.json"
        url = f"https://www.preregshortcuts.com/api/questions/get/?bank={bank_name.replace(' ', '+')}&num_questions=40&weighting={weighting}"

    Path(OUTPUT_DIR).mkdir(exist_ok=True)

    curl_command = [
        "curl", "--path-as-is", "-s", "-k", "-X", "GET",
        "-H", "Host: www.preregshortcuts.com",
        "-H", "Sec-Ch-Ua-Platform: \"macOS\"",
        "-H", "X-Csrftoken: fdwcLsSXop1Bpzupit3KItF6cDNVSJ7S",
        "-H", "Accept-Language: en-GB,en;q=0.9",
        "-H", "Accept: application/json, text/plain, */*",
        "-H", "Sec-Ch-Ua: \"Not)A;Brand\";v=\"8\", \"Chromium\";v=\"138\"",
        "-H", "User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138.0.0.0 Safari/537.36",
        "-H", "Sec-Ch-Ua-Mobile: ?0",
        "-H", "Sec-Fetch-Site: same-origin",
        "-H", "Sec-Fetch-Mode: cors",
        "-H", "Sec-Fetch-Dest: empty",
        "-H", f"Referer: https://www.preregshortcuts.com/question-bank/quiz/{bank_name.lower().replace(' ', '-')}/",
        "-H", "Accept-Encoding: gzip, deflate, br",
        "-H", "Priority: u=1, i",
        "-H", "Connection: keep-alive",
        "-b", "cookies_accepted_categories=technically_required,analytics; _ga=GA1.1.1762998342.1753816009; csrftoken=fdwcLsSXop1Bpzupit3KItF6cDNVSJ7S; sessionid=mqloafb1vcqt00hspwv4eg681kl87t4m; __stripe_mid=d458e04e-aee5-4ea2-a54e-137944451c30d2a9de; __stripe_sid=04313b41-ba0c-4c48-b748-baaf406d6083d6e8b9",
        url
    ]

    seen_hashes = set()
    unique_questions = []

    weighting_text = f" with weighting '{weighting}'" if weighting else ""

    # Load existing data if available
    if output_file.exists():
        with open(output_file, "r") as f:
            try:
                unique_questions = json.load(f)
                seen_hashes = {hash_question(q) for q in unique_questions}
                print(f"🔁 Loaded {len(unique_questions)} existing unique questions for bank '{bank_name}'{weighting_text}.")
            except json.JSONDecodeError:
                print("⚠️ Failed to load existing data — starting fresh.")

    no_new_count = 0
    batch_num = 0

    while True:
        batch_num += 1
        print(f"\n🔄 Batch {batch_num} for bank '{bank_name}'{weighting_text}...")
        batch = fetch_batch(curl_command)
        if not batch:
            break

        added = 0
        for q in batch:
            h = hash_question(q)
            if h not in seen_hashes:
                seen_hashes.add(h)
                unique_questions.append(q)
                added += 1

        print(f"✅ Added {added} unique of {len(batch)}")
        print(f"📊 Total unique questions so far for bank '{bank_name}'{weighting_text}: {len(unique_questions)}")

        if added == 0:
            no_new_count += 1
            print(f"⚠️ No new questions found. ({no_new_count}/10)")
        else:
            no_new_count = 0

        if no_new_count >= 10:
            print(f"🛑 No new questions for 10 batches for bank '{bank_name}'{weighting_text} — stopping.")
            break

        time.sleep(SLEEP_SECONDS)

    output_file.write_text(json.dumps(unique_questions, indent=2))
    print(f"\n💾 Saved {len(unique_questions)} unique questions to {output_file}")

if __name__ == "__main__":
    main()

File: Burp/grabquestions.js
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const readline = require('readline');
let fetch;
try {
  fetch = require('node-fetch');
} catch (err) {
  // fallback to built in fetch if node-fetch isn't available
  fetch = global.fetch;
}

const BANK_CHOICES = [
  'Calculations',
  'Clinical Therapeutics',
  'Clinical MEP',
  'Clinical OTC',
  'Clinical Mixed'
];

const WEIGHTING_CHOICES = ['Low', 'Medium', 'High'];

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

function ask(query) {
  return new Promise(resolve => rl.question(query, resolve));
}

async function promptForBank() {
  console.log('Select a question bank:');
  BANK_CHOICES.forEach((name, i) => console.log(`${i + 1}. ${name}`));
  while (true) {
    const ans = await ask('Enter number: ');
    const idx = parseInt(ans, 10);
    if (idx >= 1 && idx <= BANK_CHOICES.length) return BANK_CHOICES[idx - 1];
    console.log('Invalid input. Try again.');
  }
}

async function promptForWeighting() {
  console.log('Select a weighting:');
  WEIGHTING_CHOICES.forEach((name, i) => console.log(`${i + 1}. ${name}`));
  while (true) {
    const ans = await ask('Enter number: ');
    const idx = parseInt(ans, 10);
    if (idx >= 1 && idx <= WEIGHTING_CHOICES.length) return WEIGHTING_CHOICES[idx - 1];
    console.log('Invalid input. Try again.');
  }
}

function hashQuestion(q) {
  return crypto
    .createHash('sha256')
    .update((q.title || '') + (q.text || ''))
    .digest('hex');
}

async function fetchBatch(url, headers) {
  try {
    const res = await fetch(url, { headers });
    const text = await res.text();
    return JSON.parse(text);
  } catch (err) {
    console.log('❌ Failed to decode JSON');
    return [];
  }
}

async function main() {
  const bankName = await promptForBank();
  let weighting = null;
  let url;
  let outputFile;

  if (bankName === 'Calculations') {
    console.log(`\n📥 Scraping '${bankName}' questions (no weighting)...\n`);
    outputFile = path.join('question_banks', `${bankName.toLowerCase().replace(/ /g, '_')}_questions.js`);
    url = `https://www.preregshortcuts.com/api/questions/get/?bank=${bankName.replace(/ /g, '+')}&num_questions=40`;
  } else {
    weighting = await promptForWeighting();
    console.log(`\n📥 Scraping '${bankName}' questions with '${weighting}' weighting...\n`);
    outputFile = path.join('question_banks', `${bankName.toLowerCase().replace(/ /g, '_')}_${weighting.toLowerCase()}_questions.js`);
    url = `https://www.preregshortcuts.com/api/questions/get/?bank=${bankName.replace(/ /g, '+')}&num_questions=40&weighting=${weighting}`;
  }

  fs.mkdirSync('question_banks', { recursive: true });

  const headers = {
    'Host': 'www.preregshortcuts.com',
    'Sec-Ch-Ua-Platform': '"macOS"',
    'X-Csrftoken': 'fdwcLsSXop1Bpzupit3KItF6cDNVSJ7S',
    'Accept-Language': 'en-GB,en;q=0.9',
    'Accept': 'application/json, text/plain, */*',
    'Sec-Ch-Ua': '"Not)A;Brand";v="8", "Chromium";v="138"',
    'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138.0.0.0 Safari/537.36',
    'Sec-Ch-Ua-Mobile': '?0', // Browser compatibility header
    'Sec-Fetch-Site': 'same-origin',
    'Sec-Fetch-Mode': 'cors',
    'Sec-Fetch-Dest': 'empty',
    'Referer': `https://www.preregshortcuts.com/question-bank/quiz/${bankName.toLowerCase().replace(/ /g, '-')}/`,
    'Accept-Encoding': 'gzip, deflate, br',
    'Priority': 'u=1, i',
    'Connection': 'keep-alive',
    'Cookie': 'cookies_accepted_categories=technically_required,analytics; _ga=GA1.1.1762998342.1753816009; csrftoken=fdwcLsSXop1Bpzupit3KItF6cDNVSJ7S; sessionid=mqloafb1vcqt00hspwv4eg681kl87t4m; __stripe_mid=d458e04e-aee5-4ea2-a54e-137944451c30d2a9de; __stripe_sid=04313b41-ba0c-4c48-b748-baaf406d6083d6e8b9'
  };

  let uniqueQuestions = [];
  let seenHashes = new Set();
  const weightingText = weighting ? ` with weighting '${weighting}'` : '';

  if (fs.existsSync(outputFile)) {
    try {
      const code = fs.readFileSync(outputFile, 'utf8');
      const mod = await import('data:text/javascript,' + encodeURIComponent(code));
      uniqueQuestions = Array.isArray(mod.default) ? mod.default : [];
      seenHashes = new Set(uniqueQuestions.map(hashQuestion));
      console.log(`🔁 Loaded ${uniqueQuestions.length} existing unique questions for bank '${bankName}'${weightingText}.`);
    } catch (err) {
      console.log('⚠️ Failed to load existing data — starting fresh.');
    }
  }

  let noNewCount = 0;
  let batchNum = 0;

  while (true) {
    batchNum += 1;
    console.log(`\n🔄 Batch ${batchNum} for bank '${bankName}'${weightingText}...`);
    const batch = await fetchBatch(url, headers);
    if (!batch || batch.length === 0) break;

    let added = 0;
    for (const q of batch) {
      const h = hashQuestion(q);
      if (!seenHashes.has(h)) {
        seenHashes.add(h);
        uniqueQuestions.push(q);
        added += 1;
      }
    }

    console.log(`✅ Added ${added} unique of ${batch.length}`);
    console.log(`📊 Total unique questions so far for bank '${bankName}'${weightingText}: ${uniqueQuestions.length}`);

    if (added === 0) {
      noNewCount += 1;
      console.log(`⚠️ No new questions found. (${noNewCount}/10)`);
    } else {
      noNewCount = 0;
    }

    if (noNewCount >= 10) {
      console.log(`🛑 No new questions for 10 batches for bank '${bankName}'${weightingText} — stopping.`);
      break;
    }

    await new Promise(res => setTimeout(res, 2000));
  }

  fs.writeFileSync(outputFile, 'export default ' + JSON.stringify(uniqueQuestions, null, 2));
  console.log(`\n💾 Saved ${uniqueQuestions.length} unique questions to ${outputFile}`);
  rl.close();
}

main();
